File: test/knotdet.py
"""Независимый инвариант: определитель узла |Δ(−1)| (Александер) по 3D-ломаной.
Проекция вдоль случайного направления, пересечения отрезков (numpy), дуги
Виртингера, точный целочисленный определитель (Барейс)."""
import json, sys, glob, random
import numpy as np
from multiprocessing import Pool

def bareiss(M):
    n=len(M)
    if n==0: return 1
    M=[row[:] for row in M]; prev=1; sign=1
    for k in range(n-1):
        if M[k][k]==0:
            sw=next((r for r in range(k+1,n) if M[r][k]!=0), None)
            if sw is None: return 0
            M[k],M[sw]=M[sw],M[k]; sign=-sign
        for i in range(k+1,n):
            for j in range(k+1,n):
                M[i][j]=(M[i][j]*M[k][k]-M[i][k]*M[k][j])//prev
        prev=M[k][k]
    return sign*M[n-1][n-1]

def crossings(V, d):
    V=np.asarray(V,float); N=len(V)
    d=d/np.linalg.norm(d)
    a=np.array([1.0,0,0]) if abs(d[0])<0.9 else np.array([0,1.0,0])
    e1=np.cross(d,a); e1/=np.linalg.norm(e1); e2=np.cross(d,e1)
    P=np.stack([V@e1, V@e2],1); z=V@d
    I,J=np.triu_indices(N,k=2)
    m=~((I==0)&(J==N-1)); I=I[m]; J=J[m]
    A=P[I]; B=P[(I+1)%N]; C=P[J]; D=P[(J+1)%N]
    r=B-A; q=D-C; ca=C-A
    den=r[:,0]*q[:,1]-r[:,1]*q[:,0]
    ok=np.abs(den)>1e-12*(np.linalg.norm(r,axis=1)*np.linalg.norm(q,axis=1)+1e-30)
    s=np.zeros(len(I)); t=np.zeros(len(I))
    s[ok]=(ca[ok,0]*q[ok,1]-ca[ok,1]*q[ok,0])/den[ok]
    t[ok]=(ca[ok,0]*r[ok,1]-ca[ok,1]*r[ok,0])/den[ok]
    hit=ok&(s>1e-9)&(s<1-1e-9)&(t>1e-9)&(t<1-1e-9)
    out=[]
    for i,j,si,ti in zip(I[hit],J[hit],s[hit],t[hit]):
        zi=z[i]+si*(z[(i+1)%N]-z[i]); zj=z[j]+ti*(z[(j+1)%N]-z[j])
        pi=i+si; pj=j+ti
        if zi>zj: out.append((pi,pj))   # (over_pos, under_pos)
        else: out.append((pj,pi))
    return out

def det_from_crossings(cr):
    n=len(cr)
    if n==0: return 1
    under=sorted((u,c) for c,(o,u) in enumerate(cr))
    upos=[u for u,c in under]; idx_of={c:m for m,(u,c) in enumerate(under)}
    import bisect
    def arc_of(p):
        k=bisect.bisect_right(upos,p)  # число событий <= p
        return (k-1)%n
    M=[[0]*n for _ in range(n)]
    for c,(o,u) in enumerate(cr):
        m=idx_of[c]; over=arc_of(o); inn=(m-1)%n; out=m
        M[c][over]+=2; M[c][inn]-=1; M[c][out]-=1
    sub=[row[:n-1] for row in M[:n-1]]
    return abs(bareiss(sub))

def robust_det(V, ndir=7, seed=1):
    rng=random.Random(seed); vals=[]
    for k in range(ndir):
        d=np.array([rng.gauss(0,1) for _ in range(3)])
        try: vals.append(det_from_crossings(crossings(V,d)))
        except Exception as e: vals.append(None)
    ok=[v for v in vals if v is not None]
    if not ok: return {'mode':None,'agree':0,'valid':0,'vals':vals}
    mode=max(set(ok), key=ok.count)
    return {'mode':mode,'agree':ok.count(mode),'valid':len(ok),'vals':vals}

def work(args):
    i,rec,final=args
    out={'i':i,'nc':rec['nc'],'N':rec['N'],'det2d':rec['det2d'],'runDet':rec['runDet'],'runUnknot':rec['runUnknot']}
    # проекция лифта строго вдоль z (диаграмма редактора) с крошечным наклоном
    try: out['liftZ']=det_from_crossings(crossings(rec['lift'], np.array([1e-3,2e-3,1.0])))
    except Exception as e: out['liftZ']=None
    out['lift']=robust_det(rec['lift'],7,100+i)
    out['feas']=robust_det(rec['feas'],7,200+i)
    out['final']=robust_det(final,7,300+i) if final is not None else None
    return out

if __name__=='__main__':
    S=sys.argv[1]
    dumps={}
    for f in glob.glob(S+'/dump_w?.jsonl'):
        for l in open(f):
            if l.strip(): r=json.loads(l); dumps[r['i']]=r
    finals={}
    for f in glob.glob(S+'/rk_w?.verts.jsonl'):
        for l in open(f):
            if l.strip(): r=json.loads(l); finals[r['i']]=r['verts']
    jobs=[(i,dumps[i],finals.get(i)) for i in sorted(dumps)]
    with Pool(8) as p: res=p.map(work, jobs)
    json.dump(res, open(S+'/exactdet.json','w'))
    lf=sum(1 for r in res if r['lift']['mode']==r['det2d'])
    lz=sum(1 for r in res if r['liftZ']==r['det2d'])
    fp=sum(1 for r in res if r['lift']['mode'] is not None and r['feas']['mode']==r['lift']['mode'])
    dp=sum(1 for r in res if r['feas']['mode'] is not None and r['final'] and r['final']['mode']==r['feas']['mode'])
    tp=sum(1 for r in res if r['lift']['mode'] is not None and r['final'] and r['final']['mode']==r['lift']['mode'])
    ag=lambda k: sorted(r[k]['agree'] for r in res if r[k])[len(res)//2]
    print(f"n={len(res)} liftZ==det2d: {lz}  lift(robust)==det2d: {lf}  feas==lift: {fp}  final==feas: {dp}  final==lift: {tp}")
    print('agree p50: lift',ag('lift'),'feas',ag('feas'),'final',ag('final'))
    print('editor runDet == exact lift mode:', sum(1 for r in res if r['runDet']==r['lift']['mode']))
    for r in res:
        L,F,Z=r['lift'],r['feas'],r['final']
        flag=('' if L['mode']==r['det2d'] else 'LIFT≠2D ')+('' if F['mode']==L['mode'] else 'FEAS≠LIFT ')+('' if Z['mode']==F['mode'] else 'FINAL≠FEAS ')
        if flag or L['agree']<7 or F['agree']<7 or Z['agree']<7:
            print(f"  i={r['i']:2d} nc={r['nc']:2d} det2d={r['det2d']} runDet={r['runDet']} liftZ={r['liftZ']} lift={L['mode']}({L['agree']}/{L['valid']}) feas={F['mode']}({F['agree']}/{F['valid']}) final={Z['mode']}({Z['agree']}/{Z['valid']}) {flag}")

File: test/knotdet.js
// Независимый инвариант: определитель узла |Δ(−1)| по 3D-ломаной (Node, BigInt).
// Проекция вдоль случайного направления → пересечения отрезков → дуги Виртингера →
// точный целочисленный определитель (Барейс). Запуск: WORKER=i WORKERS=n node knotdet.js DIR
const fs=require('fs'), path=require('path');
function bareiss(M){ const n=M.length; if(n===0) return 1n;
  M=M.map(r=>r.map(x=>BigInt(x))); let prev=1n, sign=1n;
  for(let k=0;k<n-1;k++){
    if(M[k][k]===0n){ let sw=-1; for(let r=k+1;r<n;r++){ if(M[r][k]!==0n){ sw=r; break; } } if(sw<0) return 0n;
      const t=M[k]; M[k]=M[sw]; M[sw]=t; sign=-sign; }
    for(let i=k+1;i<n;i++) for(let j=k+1;j<n;j++) M[i][j]=(M[i][j]*M[k][k]-M[i][k]*M[k][j])/prev;
    prev=M[k][k]; }
  return sign*M[n-1][n-1]; }
function crossings(V,d){ const N=V.length; const dl=Math.hypot(d[0],d[1],d[2]); d=[d[0]/dl,d[1]/dl,d[2]/dl];
  const a=Math.abs(d[0])<0.9?[1,0,0]:[0,1,0];
  let e1=[d[1]*a[2]-d[2]*a[1], d[2]*a[0]-d[0]*a[2], d[0]*a[1]-d[1]*a[0]]; const l1=Math.hypot(...e1); e1=e1.map(x=>x/l1);
  const e2=[d[1]*e1[2]-d[2]*e1[1], d[2]*e1[0]-d[0]*e1[2], d[0]*e1[1]-d[1]*e1[0]];
  const X=new Float64Array(N), Y=new Float64Array(N), Z=new Float64Array(N);
  for(let i=0;i<N;i++){ const v=V[i]; X[i]=v[0]*e1[0]+v[1]*e1[1]+v[2]*e1[2]; Y[i]=v[0]*e2[0]+v[1]*e2[1]+v[2]*e2[2]; Z[i]=v[0]*d[0]+v[1]*d[1]+v[2]*d[2]; }
  const out=[];
  for(let i=0;i<N;i++){ const i1=(i+1)%N; const ax=X[i],ay=Y[i], rx=X[i1]-ax, ry=Y[i1]-ay; const rl=Math.hypot(rx,ry);
    for(let j=i+2;j<N;j++){ if(i===0 && j===N-1) continue; const j1=(j+1)%N;
      const cx=X[j],cy=Y[j], qx=X[j1]-cx, qy=Y[j1]-cy;
      const den=rx*qy-ry*qx; if(Math.abs(den)<=1e-12*(rl*Math.hypot(qx,qy)+1e-30)) continue;
      const cax=cx-ax, cay=cy-ay; const s=(cax*qy-cay*qx)/den; if(s<=1e-9||s>=1-1e-9) continue;
      const t=(cax*ry-cay*rx)/den; if(t<=1e-9||t>=1-1e-9) continue;
      const zi=Z[i]+s*(Z[i1]-Z[i]), zj=Z[j]+t*(Z[j1]-Z[j]); const pi=i+s, pj=j+t;
      out.push(zi>zj? [pi,pj] : [pj,pi]); } }
  return out; }
function detFromCrossings(cr){ const n=cr.length; if(n===0) return 1n;
  const under=cr.map((c,idx)=>[c[1],idx]).sort((a,b)=>a[0]-b[0]); const upos=under.map(u=>u[0]); const idxOf=new Map(under.map((u,m)=>[u[1],m]));
  const arcOf=p=>{ let lo=0,hi=n; while(lo<hi){ const mid=(lo+hi)>>1; if(upos[mid]<=p) lo=mid+1; else hi=mid; } return ((lo-1)%n+n)%n; };
  const M=[]; for(let c=0;c<n;c++) M.push(new Array(n).fill(0));
  for(let c=0;c<n;c++){ const m=idxOf.get(c), over=arcOf(cr[c][0]), inn=((m-1)%n+n)%n, out=m; M[c][over]+=2; M[c][inn]-=1; M[c][out]-=1; }
  const sub=M.slice(0,n-1).map(r=>r.slice(0,n-1)); const d=bareiss(sub); return d<0n?-d:d; }
function rng(seed){ let s=seed>>>0; return ()=>{ s=(s*1664525+1013904223)>>>0; return s/4294967296; }; }
function gauss(r){ const u=r()||1e-12, v=r(); return Math.sqrt(-2*Math.log(u))*Math.cos(2*Math.PI*v); }
function robust(V, ndir, seed){ const r=rng(seed); const vals=[];
  for(let k=0;k<ndir;k++){ const d=[gauss(r),gauss(r),gauss(r)]; try{ vals.push(detFromCrossings(crossings(V,d)).toString()); }catch(e){ vals.push(null); } }
  const ok=vals.filter(x=>x!==null); const cnt={}; for(const x of ok) cnt[x]=(cnt[x]||0)+1;
  let mode=null, mc=0; for(const k in cnt) if(cnt[k]>mc){ mc=cnt[k]; mode=k; }
  return {mode, agree:mc, valid:ok.length, vals}; }
const DIR=process.argv[2], W=+(process.env.WORKER||0), NW=+(process.env.WORKERS||1);
const dumps={}, finals={};
for(const f of fs.readdirSync(DIR)){ if(/^dump_w\d\.jsonl$/.test(f)) for(const l of fs.readFileSync(path.join(DIR,f),'utf8').split('\n')){ if(l.trim()){ const r=JSON.parse(l); dumps[r.i]=r; } }
  if(/^rk_w\d\.verts\.jsonl$/.test(f)) for(const l of fs.readFileSync(path.join(DIR,f),'utf8').split('\n')){ if(l.trim()){ const r=JSON.parse(l); finals[r.i]=r.verts; } } }
const res=[];
for(const i of Object.keys(dumps).map(Number).sort((a,b)=>a-b)){ if(i%NW!==W) continue; const rec=dumps[i];
  const out={i, nc:rec.nc, N:rec.N, det2d:rec.det2d, runDet:rec.runDet, runUnknot:rec.runUnknot};
  try{ out.liftZ=detFromCrossings(crossings(rec.lift,[1e-3,2e-3,1])).toString(); }catch(e){ out.liftZ=null; }
  out.lift=robust(rec.lift,7,100+i); out.feas=robust(rec.feas,7,200+i); out.final=finals[i]? robust(finals[i],7,300+i) : null;
  res.push(out); }
fs.writeFileSync(path.join(DIR,'exact_w'+W+'.json'), JSON.stringify(res));
console.log('worker',W,'done',res.length);
